core/project_settings.h

```diff
@@ -216,6 +220,22 @@ Variant _GLOBAL_DEF_ALIAS(const String &p_var, const String &p_old_name, const V
 #define GLOBAL_DEF_RST_NOVAL(m_var, m_value) _GLOBAL_DEF(m_var, m_value, true, true)
 #define GLOBAL_DEF_ALIAS(m_var, m_old_name, m_value) _GLOBAL_DEF_ALIAS(m_var, m_old_name, m_value)
 #define GLOBAL_DEF_ALIAS_RST(m_var, m_old_name, m_value) _GLOBAL_DEF(m_var, m_old_name, m_value, true)
-#define GLOBAL_GET(m_var) ProjectSettings::get_singleton()->get(m_var)
+
+// Uncached, slow, but has no static variables.
+#define GLOBAL_GET_UNCACHED(m_var) ProjectSettings::get_singleton()->get(m_var)
+
+// Cached but uses a typed variable for storage, this can be more efficient.
+#define GLOBAL_GET_FAST(VAR_TYPE, SETTING_NAME) ([=]() -> const VAR_TYPE & {\
+	static VAR_TYPE local_var;                                                   \
+		static uint32_t local_version = 0;                                      \
+		uint32_t new_version = ProjectSettings::get_singleton()->get_version(); \
+		if (local_version != new_version) {                                     \
+			local_version = new_version;                                        \
+			local_var = ProjectSettings::get_singleton()->get(SETTING_NAME);     \
+		}					                                                   \
+		return local_var; })()
```

Comment:
> In particular, I had to specify a default capture mode to get it to compile for some parts that were using GLOBAL_GET.

You can pass `SETTING_NAME` in as a parameter to avoid capturing anything.

```suggestion
#define GLOBAL_GET_FAST(VAR_TYPE, SETTING_NAME) ([](const StringName &p_name) -> const VAR_TYPE & {\
	static VAR_TYPE local_var;                                                   \
		static uint32_t local_version = 0;                                      \
		uint32_t new_version = ProjectSettings::get_singleton()->get_version(); \
		if (local_version != new_version) {                                     \
			local_version = new_version;                                        \
			local_var = ProjectSettings::get_singleton()->get(p_name);     \
		}					                                                   \
		return local_var; })(SETTING_NAME)
```

Also maybe this macro should be `GLOBAL_GET_CACHED` since `GLOBAL_GET_UNCACHED` is used?

Reply:
Ah excellent I'll try that with the parameter. :+1: 

On the naming, I'm open to suggestions. When I talked with @akien-mga  earlier we were going with the idea of replacing the existing `GLOBAL_GET` wholesale for most cases. The uncached version I added for "backward" support, but it isn't used anywhere so could potentially be removed.

So that mainly leaves what to call the `FAST` version. I did think about `GLOBAL_GET_CACHED` but, both are cached now, so the only difference is that this version stores a type instead of a variant. It could be called e.g. `GLOBAL_GET_TYPED`, but then that isn't very clear from the use areas _why_ it might be used in preference to `GLOBAL_GET`. But I'm welcoming any ideas here.